src/number_theory/number_field/embeddings.lean

```diff
@@ -106,20 +93,17 @@ open finite_dimensional polynomial set
 variables {K : Type*} [field K] [number_field K]
 variables {A : Type*} [normed_field A] [is_alg_closed A] [normed_algebra ℚ A]
 
-lemma coeff_bdd_of_norm_le {B : ℝ} {x : K} (h : ∀ φ : K →+* A, ∥φ x∥ ≤ B) (i : ℕ):
+lemma coeff_bdd_of_norm_le {B : ℝ} {x : K} (h : ∀ φ : K →+* A, ∥φ x∥ ≤ B) (i : ℕ) :
   ∥(minpoly ℚ x).coeff i∥ ≤ (max B 1) ^ (finrank ℚ K) * (finrank ℚ K).choose ((finrank ℚ K) / 2) :=
 begin
   have hx : is_integral ℚ x := is_separable.is_integral _ _,
-  rw (by rw [coeff_map, norm_algebra_map'] :
-    ∥(minpoly ℚ x).coeff i∥ = ∥(map (algebra_map ℚ A) (minpoly ℚ x)).coeff i∥),
+  rw [← norm_algebra_map' A, ← coeff_map (algebra_map ℚ A)],
   refine coeff_bdd_of_roots_le _ (minpoly.monic hx) (is_alg_closed.splits_codomain _)
-    (minpoly.nat_degree_le hx) _ i,
-  intros z hz,
-  rsuffices ⟨φ, rfl⟩ : ∃ (φ : K →+* A), φ x = z, {exact h φ },
-  letI : char_zero A := char_zero_of_injective_algebra_map (algebra_map ℚ _).injective,
-  rw [← set.mem_range, rat_range_eq_roots, mem_root_set_iff, aeval_def],
-  convert (mem_roots_map _).mp hz,
-  repeat { exact monic.ne_zero (minpoly.monic hx), },
+    (minpoly.nat_degree_le hx) (λ z hz, _) i,
+  classical, rw [← multiset.mem_to_finset, ← finset.mem_coe] at hz,
+  haveI : char_zero A := char_zero_of_injective_algebra_map (algebra_map ℚ A).injective,
+  obtain ⟨φ, rfl⟩ := (rat_range_eq_roots K A x).symm.subset (_ : z ∈ _), { exact h φ },
+  rw root_set_def, convert hz using 5,
```

Comment:
This is a slow `convert` and the whole proof is slightly shorter but slower than the original one, which also uses `convert`. Is this a known diamond (related to `char_zero`?) that's being fixed?

Reply:
cc @ericrbg @Vierkantor

Reply:
Oh I see: there's a `normed_algebra ℚ A` instance in the `variables` line above, which is definitely going to cause diamonds with the default `algebra_rat`.

Reply:
My conclusion is: either we look for/define a class on `A` itself that together with `normed_field A` implies `normed_algebra ℚ A` (like `char_zero A` does for `algebra ℚ A`), or we keep this `convert`.

Reply:
Thanks for the diagnosis! My solution is to replace the argument `[char_zero A]` of `rat_range_eq_roots` (now renamed `range_eval_eq_root_set_minpoly`) by `[algebra ℚ A]`, which is more general, since we automatically get `[algebra ℚ A]` from `[char_zero A]` and `algebra_rat` because we also have `[field A]` around. Then I golf down this proof further using the restored defeq.